openshift/odo
# add state to ingress list and describe

Signed-off-by: Stephanie <stephanie.cao@ibm.com>

**What type of PR is this?**
> Uncomment only one ` /kind` line, and delete the rest.
> For example, `> /kind bug` would simply become: `/kind bug`

/area url
/kind feature
/area devfile

>
> Documentation changes: Please include [skip ci] in your commit message as well
> /kind documentation
> [skip ci]

**What does does this PR do / why we need it**:
This PR adds `state` for ingresses when running`url list` and `url describe`. 
This PR also modifies the existing integration tests since the expected output has been change for `url list` and `url describe`

**Which issue(s) this PR fixes**:

Fixes #2787 
Fixes https://github.com/openshift/odo/issues/2983
Fixes #3060

**How to test changes / Special notes to the reviewer**:
Automation Tests: 
simply run `make test-cmd-devfile-url`
```
$ make test-cmd-devfile-url
ginkgo  -randomizeAllSpecs -slowSpecThreshold=120 -timeout 7200s -nodes=2 -focus="odo devfile url command tests" tests/integration/devfile/
Running Suite: Devfile Suite
============================
Random Seed: 1589310561 - Will randomize all specs
Will run 53 specs

Running in parallel across 2 nodes

SSSSSSSSSSSSSSSSSSSSSSSS•SS•SSSSS•SS•SSSSSSSSSSS•S•••
Ran 8 of 53 Specs in 223.524 seconds
SUCCESS! -- 8 Passed | 0 Failed | 0 Pending | 45 Skipped


Ginkgo ran 1 suite in 3m46.272474203s
Test Suite Passed
```

Manual Tests
```
odo preference set experimental true
odo create java-spring-boot
odo url create ingressurl1 --host 1.2.3.4.com --ingress --secure
```
Run `odo url list` should see the state of the url is `Not Pushed`
```
$ odo url list
Found the following URLs for component java-spring-boot
NAME            STATE          URL                                 PORT     SECURE
ingressurl1     Not Pushed     https://ingressurl1.1.2.3.4.com     8080     true

There are local changes. Please run 'odo push'.
```
Run `odo url list -o json` should see the json format
```
odo url list -o json
{
    "kind": "List",
    "apiVersion": "odo.dev/v1alpha1",
    "metadata": {},
    "items": [
        {
            "kind": "url",
            "apiVersion": "odo.dev/v1alpha1",
            "metadata": {
                "name": "ingressurl1",
                "creationTimestamp": null
            },
            "spec": {
                "tls": [
                    {
                        "hosts": [
                            "ingressurl1.1.2.3.4.com"
                        ],
                        "secretName": "java-spring-boot-tlssecret"
                    }
                ],
                "rules": [
                    {
                        "host": "ingressurl1.1.2.3.4.com",
                        "http": {
                            "paths": [
                                {
                                    "path": "/",
                                    "backend": {
                                        "serviceName": "java-spring-boot",
                                        "servicePort": 8080
                                    }
                                }
                            ]
                        }
                    }
                ]
            },
            "status": {
                "state": "Not Pushed"
            }
        }
    ]
}

```
run `odo push`, and then list the url, should see state as `Pushed`
```
$ odo url list
NAME            STATE      URL                                 PORT     SECURE
ingressurl1     Pushed     https://ingressurl1.1.2.3.4.com     8080     true
```

run `odo url delete ingressurl1`, and then list the url, should see state as `Locally Deleted`
```
$ odo url delete ingressurl1
? Are you sure you want to delete the url ingressurl1 Yes
 ✓  URL ingressurl1 removed from the env file

To delete the URL on the cluster, please use `odo push`
$ odo url list
Found the following URLs for component java-spring-boot
NAME            STATE               URL                                 PORT     SECURE
ingressurl1     Locally Deleted     https://ingressurl1.1.2.3.4.com     8080     true
```

Manual tests for describe are similar.
